<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.506">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Leo Lahti">

<title>Tabular data analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    <div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="day2.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Tabular data analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Leo Lahti </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="recap-of-day-1" class="level2">
<h2 class="anchored" data-anchor-id="recap-of-day-1">Recap of Day 1</h2>
<ul>
<li><strong>Data</strong>: <em>SummarizedExperiment</em>: rowData, colData, assays</li>
<li><strong>Methods</strong>: subsetting, transformations</li>
</ul>
<p><img src="images/paste-AA8BAE15.png" class="img-fluid" width="402"></p>
</section>
<section id="day-1-basic-data-wrangling" class="level2">
<h2 class="anchored" data-anchor-id="day-1-basic-data-wrangling">Day 1: Basic data wrangling</h2>
<ul>
<li>reproducible data science workflow</li>
<li>data import</li>
<li>data containers</li>
<li>data manipulation (subsetting, transformations)</li>
</ul>
</section>
<section id="todays-learning-goals" class="level2">
<h2 class="anchored" data-anchor-id="todays-learning-goals">Today’s learning goals</h2>
<p>Expanding multi-assay analyses:<br>
<strong>TreeSummarizedExperiment</strong></p>
<ul>
<li>augmenting the data (add diversities)</li>
<li>data agglomeration &amp; <em>alternative experiments</em></li>
<li>tree information: <em>rowTree, colTree</em></li>
</ul>
</section>
<section id="todays-program" class="level2">
<h2 class="anchored" data-anchor-id="todays-program">Today’s program</h2>
<p>Morning: data wrangling</p>
<p>Afternoon: data visualizations</p>
</section>
<section id="data-enrichment" class="level1">
<h1>Data enrichment</h1>
<section id="visualizing-coldata" class="level2">
<h2 class="anchored" data-anchor-id="visualizing-coldata">Visualizing colData</h2>
<p>Task: visualize the abundance of a specific microbial Species against the measurement Site</p>
</section>
<section id="alpha-diversity-task" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity-task">Alpha diversity task</h2>
<p>Use the available tools to assess and visualize alpha diversity, and augment colData</p>
<ul>
<li>Exercises 17.5.1-17.5.2</li>
<li>Add Shannon diversity in colData</li>
<li>Visualize diversity differences between sample groups</li>
</ul>
</section>
<section id="alpha-diversity-aging" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity-aging">Alpha diversity &amp; aging</h2>
<p>Healthy &amp; normal obese subjects.</p>
<p><img src="images/paste-CC6290A9.png" class="img-fluid"></p>
</section>
<section id="alpha-diversity-and-diet" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="alpha-diversity-and-diet">Alpha diversity and diet</h2>
<p><img src="images/paste-C67E84BB.png" class="img-fluid"></p>
<p><img src="images/paste-85FC6DB9.png" class="img-fluid"></p>
</section>
<section id="alpha-diversity" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity">Alpha diversity</h2>
<ul>
<li><p>How many types?</p></li>
<li><p>Distribution of types?</p></li>
<li><p>Dominance of types?</p></li>
</ul>
<p><img src="images/paste-37918533.png" class="img-fluid"></p>
</section>
<section id="alpha-diversity-1" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity-1">Alpha diversity</h2>
<ul>
<li><p>How many types?</p></li>
<li><p>Distribution of types?</p></li>
<li><p>Dominance of types?</p></li>
</ul>
<p><img src="images/paste-D240B56A.png" class="img-fluid"></p>
</section>
<section id="alpha-diversity-indices" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="alpha-diversity-indices">Alpha diversity indices</h2>
<p><strong>Richness</strong></p>
<ul>
<li><p>number of types</p></li>
<li><p>Eetimates of true richness based on finite sample sizes (Howard Sanders 1968); see e.g.&nbsp;Chao1</p></li>
</ul>
<p><strong>Evenness</strong></p>
<ul>
<li>distribution of sizes (even or uneven?)</li>
</ul>
<p><strong>Diversity</strong></p>
<ul>
<li>Combining richness &amp; evenness</li>
</ul>
<p><strong>Dominance</strong></p>
</section>
<section id="finite-sampling" class="level2">
<h2 class="anchored" data-anchor-id="finite-sampling">Finite sampling</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-178A9470.png" class="img-fluid figure-img"></p>
<figcaption>https://github.com/mblstamps/stamps2019/blob/master/STAMPS2019_overview_Pop.pdf</figcaption>
</figure>
</div>
</section>
<section id="section" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="section"></h2>
<p><img src="images/paste-B5BEB852.png" class="img-fluid"></p>
<div class="columns">
<div class="column" style="width:70%;">
<p><img src="images/paste-99C3A18C.png" class="img-fluid"></p>
</div><div class="column" style="width:30%;">
<p><img src="images/paste-E3910F24.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="section-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="section-1"></h2>
<p>High-quality reference genomes are required for functional characterization and taxonomic assignment of the human gut microbiota.</p>
<p>Unified Human Gastrointestinal Genome (UHGG):</p>
<ul>
<li><p>4,644 gut prokaryotes (&gt;70% lack cultured representatives)</p></li>
<li><p>204,938 nonredundant genomes</p></li>
<li><p>Encode &gt;170 million protein sequences, collated into Unified Human Gastrointestinal Protein (UHGP) catalog.</p></li>
</ul>
<p>UHGP more than doubles the number of gut proteins in comparison to those present in the Integrated Gene Catalog.</p>
<ul>
<li><p>40% of the UHGP lack functional annotations</p></li>
<li><p>Intraspecies genomic variation analyses revealed a large reservoir of accessory genes and single-nucleotide variants, many of which are specific to individual human populations.</p></li>
</ul>
<p>The UHGG and UHGP collections enable studies linking genotypes to phenotypes in the human gut microbiome.</p>
</section>
<section id="estimating-species-content" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="estimating-species-content">Estimating species content</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-2AD9955A.png" class="img-fluid figure-img"></p>
<figcaption>Copyright © Claudia Zirion, Diego Garfias, Vanessa Arellano, Aaron Jaime, Abel Lovaco, Daniel Díaz, Abraham Avelar, Nelly Sélem https://carpentries-incubator.github.io/metagenomics-workshop/)</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div class="rows">
<div class="row" height="70%">
<p><img src="images/paste-67F9C0C6.png" class="img-fluid"></p>
</div>
<div class="row" height="30%">
<p><img src="images/paste-B11EAEC2.png" class="img-fluid"></p>
</div>
</div>
</div>
</div>
</section>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"></h2>
<div class="columns">
<div class="column" style="width:70%;">
<p><img src="images/paste-965269D9.png" class="img-fluid"></p>
</div><div class="column" style="width:30%;">
<p><img src="images/paste-422AB61F.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="section-3" class="level2">
<h2 class="anchored" data-anchor-id="section-3"></h2>
<p><img src="images/paste-863D0ABE.png" class="img-fluid"></p>
<div class="columns">
<div class="column" style="width:80%;">
<p><img src="images/paste-29F79604.png" class="img-fluid"></p>
</div><div class="column" style="width:20%;">
<p><img src="images/paste-FFF6AC22.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="common-alpha-diversity-indices" class="level2">
<h2 class="anchored" data-anchor-id="common-alpha-diversity-indices">Common alpha diversity indices</h2>
<p><strong>Phylogenetically neutral diversities:</strong></p>
<ul>
<li>Richness (observed, Chao1, ACE)</li>
<li>Evenness (Pielou’s evenness)</li>
<li>Diversity (inverse Simpson, Shannon)</li>
</ul>
<p><strong>Phylogeny-aware diversities:</strong></p>
<ul>
<li>Faith diversity index</li>
</ul>
</section>
<section id="phylogenetic-diversity-indices" class="level2">
<h2 class="anchored" data-anchor-id="phylogenetic-diversity-indices">Phylogenetic diversity indices</h2>
<p><img src="images/paste-78D6AFB5.png" class="img-fluid"></p>
<p><img src="images/paste-DD6C75BB.png" class="img-fluid"></p>
</section>
<section id="inverse-simpson" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="inverse-simpson">Inverse Simpson</h2>
<p><img src="images/paste-1B870B03.png" class="img-fluid"></p>
<p>How likely it is to pick two members of the same species at random?</p>
<p><img src="images/paste-37918533.png" class="img-fluid"></p>
</section>
<section id="inverse-simpson-1" class="level2">
<h2 class="anchored" data-anchor-id="inverse-simpson-1">Inverse Simpson</h2>
<p><strong>Beware</strong> the variants:</p>
<ul>
<li><p>Simpson (<span class="math inline">\(\lambda\)</span>)</p></li>
<li><p>reciprocal Simpson (<span class="math inline">\(1-\lambda\)</span>)</p></li>
<li><p>inverse Simpson (<span class="math inline">\(\frac{1}{\lambda}\)</span>)</p></li>
</ul>
</section>
<section id="shannon-diversity" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="shannon-diversity">Shannon diversity</h2>
<p>Shannon Index:</p>
<p><img src="images/paste-9EC22D90.png" class="img-fluid"></p>
<p>True Richness:</p>
<p><img src="images/paste-01CE5752.png" class="img-fluid"></p>
<p><em>True diversity, or the effective number of types, refers to the number of equally abundant types needed for the average proportional abundance of the types to equal what is observed in the dataset of interest.</em></p>
</section>
<section id="evenness" class="level2">
<h2 class="anchored" data-anchor-id="evenness">Evenness</h2>
<p>H / ln(S)</p>
<ul>
<li>H: Shannon diversity</li>
<li>S: Species richness</li>
</ul>
</section>
<section id="hills-diversity-as-a-unifying-concept" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="hills-diversity-as-a-unifying-concept">Hill’s Diversity as a unifying concept</h2>
<span class="math display">\[\begin{equation}
^qD = (\sum_i^R p_i^q )^\frac{1}{1-q}
\end{equation}\]</span>
<p><strong>Hill’s alpha diversities</strong></p>
<p>R: richness (number of distinct types)</p>
<p>pi: proportion of type I</p>
<p>Order of diversity:</p>
<ul>
<li>q = 0 : Species Richness</li>
<li>q = 1 : Shannon diversity</li>
<li>q = 2 : (Inverse) Simpson diversity</li>
<li>q ≠ 1 : Renyi entropy</li>
</ul>
</section>
<section id="hills-diversity-as-a-unifying-concept-1" class="level2">
<h2 class="anchored" data-anchor-id="hills-diversity-as-a-unifying-concept-1">Hill’s Diversity as a unifying concept</h2>
<div class="columns">
<div class="column" style="width:40%;">
<p><img src="images/paste-BD993C0E.png" class="img-fluid"></p>
</div><div class="column" style="width:60%;">
<p><strong>Hill’s alpha diversities</strong></p>
<ul>
<li>Richness</li>
<li>inverse Simpson</li>
<li>Shannon</li>
</ul>
</div>
</div>
</section>
</section>
<section id="data-wrangling" class="level1">
<h1>Data wrangling</h1>
<!--
## Overview so far

By now, you are supposed to be able to:

-   understand the basic structure of the TreeSE data container

-   construct a new TreeSE object from scratch or from imported data

-   extract specific components from the object (assays, sample & feature info, trees)

-\> How to manipulate & operate with this data object?

-->
<section id="basic-data-operations" class="level2">
<h2 class="anchored" data-anchor-id="basic-data-operations">Basic data operations</h2>
<ul>
<li><p>Transform</p></li>
<li><p>Subset</p></li>
<li><p>Merge</p></li>
<li><p><strong>Aggregate</strong></p></li>
<li><p>Split</p></li>
</ul>
</section>
<section id="subsetting" class="level2">
<h2 class="anchored" data-anchor-id="subsetting">Subsetting</h2>
<p>Load example data set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SingleCellExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: TreeSummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biostrings</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: XVector</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biostrings'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    strsplit</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MultiAssayExperiment</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check dimension:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19216    26</code></pre>
</div>
</div>
<p>Check dimension for a subset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10  3</code></pre>
</div>
</div>
</section>
<section id="transformations" class="level2">
<h2 class="anchored" data-anchor-id="transformations">Transformations</h2>
<ul>
<li>Presence/absence</li>
<li>Compositional (percentages)</li>
<li><span class="math inline">\(Log_{10}\)</span></li>
<li>CLR and other <em>Aitchison</em> transformations</li>
<li>Phylogenetic transformations (e.g.&nbsp;philr)</li>
<li>Custom transformations</li>
</ul>
</section>
<section id="transformations-1" class="level2">
<h2 class="anchored" data-anchor-id="transformations-1">Transformations</h2>
<p>Task: Alternative assays</p>
<ul>
<li>visualize transformed data; histograms, boxplots</li>
<li>compare different transformations (scatterplot?)</li>
</ul>
</section>
<section id="agglomeration" class="level2">
<h2 class="anchored" data-anchor-id="agglomeration">Agglomeration</h2>
<ul>
<li>taxonomic units</li>
<li>TreeSE objects</li>
</ul>
</section>
<section id="agglomeration-1" class="level2">
<h2 class="anchored" data-anchor-id="agglomeration-1">Agglomeration</h2>
<p>Agglomerate microbiota data to higher taxonomic levels:</p>
<ul>
<li>chapter 6.3</li>
<li>agglomerateByRank</li>
<li>compare diversity or prevalent features between levels</li>
</ul>
</section>
<section id="alternative-experiments" class="level2">
<h2 class="anchored" data-anchor-id="alternative-experiments">Alternative experiments</h2>
<p>Alternative assays vs.&nbsp;alternative experiments?</p>
<ul>
<li>Store agglomerated data: <em>altExp</em></li>
<li>Do all levels at once: <em>splitByRanks</em></li>
</ul>
</section>
<section id="splits" class="level2">
<h2 class="anchored" data-anchor-id="splits">Splits</h2>
<p>Splitting by:</p>
<ul>
<li>taxonomic units</li>
<li>sample or feature groups</li>
</ul>
</section>
<section id="taxonomic-ranks-altexp" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="taxonomic-ranks-altexp">Taxonomic ranks &amp; <em>altExp</em></h2>
<p>The alternative experiments (<em>altExp</em>) mechanism allows us to include multiple abundance tables at different taxonomic levels.</p>
<table class="table">
<colgroup>
<col style="width: 17%">
<col style="width: 24%">
<col style="width: 22%">
<col style="width: 35%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: right;">Option</th>
<th style="text-align: left;">Rows (features)</th>
<th style="text-align: left;">Cols (samples)</th>
<th style="text-align: left;">Recommendation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">assays</td>
<td style="text-align: left;">match</td>
<td style="text-align: left;">match</td>
<td style="text-align: left;">Data transformations</td>
</tr>
<tr class="even">
<td style="text-align: right;"><strong>altExp</strong></td>
<td style="text-align: left;">free</td>
<td style="text-align: left;">match</td>
<td style="text-align: left;">Alternative experiments</td>
</tr>
<tr class="odd">
<td style="text-align: right;">MultiAssay</td>
<td style="text-align: left;">free</td>
<td style="text-align: left;">free (mapping)</td>
<td style="text-align: left;">Multi-omic experiments</td>
</tr>
</tbody>
</table>
</section>
<section id="alternative-experiments-and-assays" class="level2">
<h2 class="anchored" data-anchor-id="alternative-experiments-and-assays">Alternative experiments and assays?</h2>
<ul>
<li>Pick clr assay from Genus-level data table?</li>
<li>Compare Shannon diversity from Genus and Species levels?</li>
</ul>
</section>
<section id="treesummarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="treesummarizedexperiment">TreeSummarizedExperiment</h2>
<p>Huang et al.&nbsp;F1000, 2021</p>
<p><img src="images/paste-14DB8F76.png" class="img-fluid"></p>
</section>
</section>
<section id="visualization" class="level1">
<h1>Visualization</h1>
<section id="ordination" class="level2">
<h2 class="anchored" data-anchor-id="ordination">Ordination</h2>
<ul>
<li>Visualize example data with PCoA using Bray-Curtis dissimilarity</li>
<li>Visualize example data with PCA using Aitchison distance (CLR + Euclid)</li>
</ul>
</section>
<section id="heatmaps" class="level2">
<h2 class="anchored" data-anchor-id="heatmaps">Heatmaps</h2>
<ul>
<li>Visualize abundance variation for selected taxa on a heatmap</li>
</ul>
</section>
<section id="trees" class="level2">
<h2 class="anchored" data-anchor-id="trees">Trees</h2>
<ul>
<li>Visualize phylogenetic tree using the examples</li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>