<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.506">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Leo Lahti">

<title>TreeSummarizedExperiment</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    <div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="treese.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">TreeSummarizedExperiment</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Leo Lahti </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="summarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="summarizedexperiment">SummarizedExperiment</h2>
<p><img src="images/paste-AA8BAE15.png" class="img-fluid"></p>
</section>
<section id="treesummarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="treesummarizedexperiment">TreeSummarizedExperiment</h2>
<p>Huang et al.&nbsp;F1000, 2021</p>
<p><img src="images/paste-14DB8F76.png" class="img-fluid"></p>
</section>
<section id="alternative-data-container-phyloseq" class="level2">
<h2 class="anchored" data-anchor-id="alternative-data-container-phyloseq">Alternative data container: <em>phyloseq</em></h2>
<p>Current standard for (16S) microbiome bioinformatics in R (J McMurdie, S Holmes et al.)</p>
<p><img src="images/paste-43F123A2.png" class="img-fluid"></p>
</section>
<section id="open-microbiome-data" class="level2">
<h2 class="anchored" data-anchor-id="open-microbiome-data">Open microbiome data</h2>
<p>Available in TreeSummarizedExperiment format:</p>
<ul>
<li>EBI Mgnify database: MGnifyR package</li>
<li>Human studies: <a href="https://waldronlab.io/curatedMetagenomicData/articles/available-studies.html">curatedMetagenomicData</a> (Pasolli et al <a href="https://doi.org/10.1038/nmeth.4468">Nat Meth 2017</a>)</li>
<li>Other studies: <a href="https://microbiome.github.io/microbiomeDataSets/index.html">microbiomeDataSets</a> (Lahti et al.)</li>
<li>R package data (<a href="https://microbiome.github.io/mia/reference/index.html">mia</a>, <a href="https://microbiome.github.io/miaViz/reference/index.html">miaViz</a>, <a href="https://microbiome.github.io/miaTime/reference/index.html">miaTime</a>)</li>
</ul>
</section>
<section id="demo" class="level2">
<h2 class="anchored" data-anchor-id="demo">Demo</h2>
<ul>
<li><p>We have already loaded HintikkaXOData</p></li>
<li><p>Now, let us look at the data components</p></li>
</ul>
</section>
<section id="task" class="level2">
<h2 class="anchored" data-anchor-id="task">Task</h2>
<p>Explore TreeSE components (OMA 17.2); extract and visualize:</p>
<ul>
<li><p>(assays, colData, rowData)</p></li>
<li><p>rowTree, colTree</p></li>
<li><p>metadata</p></li>
</ul>
</section>
<section id="data-import" class="level1">
<h1>Data import</h1>
<section id="data-import-from-source-files" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="data-import-from-source-files">Data import from source files</h2>
<ul>
<li><p>Construct a new TreeSE object from scratch, starting from original data files.</p></li>
<li><p>Taxonomic profiling from 40 rat Cecum samples including 12706 OTUs from 318 species</p></li>
<li><p>Diet comparison with High/Low fat diet and xylo-oligosaccaride supplementation.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-A312439C.png" class="img-fluid figure-img"></p>
<figcaption>International Journal of Environmental Research and Public Health 18(8):4049 <a href="https://doi.org/10.3390/ijerph18084049" class="uri">https://doi.org/10.3390/ijerph18084049</a></figcaption>
</figure>
</div>
</section>
<section id="data-import-from-source-files-1" class="level2">
<h2 class="anchored" data-anchor-id="data-import-from-source-files-1">Data import from source files</h2>
<p>Construct a new TreeSE object from scratch, starting from original data files.</p>
<p><a href="https://microbiome.github.io/OMA/exercises.html">OMA, Chapter 18</a> (18.3.1-3) includes additional tips</p>
<ul>
<li><p>Read in the CSV files. You can use shared example data files on the cloud server; see in R: <code>dir("shared/data")</code></p></li>
<li><p>Data files include sample data (<code>coldata.csv</code>); taxonomic table (<code>rowdata_taxa.csv</code>); taxonomic abundance table (<code>assay_taxa.csv</code>). Load these in your RStudio session with e.g.&nbsp;<code>read.csv("shared/data/coldata.csv")</code></p></li>
<li><p>Construct TreeSE in R (see <a href="https://microbiome.github.io/OMA/containers.html#loading-experimental-microbiome-data">OMA Ch. 2</a>)</p></li>
</ul>
</section>
<section id="data-import-from-other-source-formats" class="level2">
<h2 class="anchored" data-anchor-id="data-import-from-other-source-formats">Data import from other source formats</h2>
<p><a href="https://microbiome.github.io/OMA/exercises.html">OMA, Chapter 18</a> (18.3.4)</p>
<ul>
<li><p>Follow the biom file example in <a href="https://microbiome.github.io/OMA/containers.html#biom-example">OMA 2.3.2.1</a></p></li>
<li><p>Example data files are available on the cloud server; see in R: <code>dir("shared/data")</code></p></li>
</ul>
</section>
<section id="data-conversions" class="level2">
<h2 class="anchored" data-anchor-id="data-conversions">Data conversions</h2>
<p><a href="https://microbiome.github.io/OMA/exercises.html">OMA, Chapter 18</a> (18.3.5)</p>
<p><em>TreeSummarizedExperiment</em> and <em>phyloseq</em> are alternative containers for microbiome data in R. It is useful to know how to convert between these two formats.</p>
<ul>
<li><p>Convert your TreeSE into phyloseq</p></li>
<li><p>Convert the phyloseq back to TreeSE</p></li>
</ul>
</section>
<section id="data-exploration" class="level2">
<h2 class="anchored" data-anchor-id="data-exploration">Data exploration</h2>
<p><a href="https://microbiome.github.io/OMA/exercises.html">OMA, Chapter 18</a> (18.4.3 - 18.4.5)</p>
<ul>
<li>Optional: if time will allow, you can try out these as well</li>
</ul>
</section>
<section id="task-1" class="level2">
<h2 class="anchored" data-anchor-id="task-1">Task</h2>
<ul>
<li><p>TreeSE/MAE, exploration &amp; analysis, how to find more</p></li>
<li><p>mia &amp; miaViz R packages:</p></li>
<li><p>Intro to miaverse (mia, miaViz, miaTime, miaSim, OMA)</p></li>
<li><p>2-3 simple examples to summarize &amp; visualize data from TreeSE container</p></li>
</ul>
</section>
</section>
<section id="assays-transformations" class="level1 scrollable">
<h1 class="scrollable">Assays &amp; transformations</h1>
<p><img src="images/paste-092DD78B.png" class="img-fluid"></p>
<p><img src="images/paste-65D4EBAC.png" class="img-fluid"></p>
<section id="normalizing-library-size" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="normalizing-library-size">Normalizing library size?</h2>
<p>Bias in compositional data:</p>
<ul>
<li>If sample A has been sampled deeper than sample B, also the counts can be expected to be higher.</li>
</ul>
<p>Possible solutions:</p>
<ul>
<li><p>Divide by the total number of reads per sample (<em>compositional abundance</em>)</p></li>
<li><p>Rarify (subsample) to even sampling depth</p></li>
</ul>
<p>→ Problem: <strong>Abundant taxa may distort the ratios</strong></p>
<p><img src="images/paste-D873776C.png" class="img-fluid"></p>
</section>
<section id="relative-vs.-absolute-abundances" class="level2">
<h2 class="anchored" data-anchor-id="relative-vs.-absolute-abundances">Relative vs.&nbsp;absolute abundances</h2>
<p>-&gt; Compositional data analysis (CoDa)</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-F928D6AC.png" class="img-fluid figure-img"></p>
<figcaption>Vandeputte et al.&nbsp;Nature 551:507-511, 2017</figcaption>
</figure>
</div>
</section>
<section id="rmp-vs.-qmp" class="level2">
<h2 class="anchored" data-anchor-id="rmp-vs.-qmp">RMP vs.&nbsp;QMP</h2>
<p><strong>Potentially drastic effect on conclusions!</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-B5DEEE10.png" class="img-fluid figure-img"></p>
<figcaption>Vandeputte et al.&nbsp;Nature 551:507-511, 2017</figcaption>
</figure>
</div>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"></h2>
<p><strong>Abundance along the community landscape</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-A55C75AB.png" class="img-fluid figure-img"></p>
<figcaption>Vandeputte et al.&nbsp;Nature 551:507-511, 2017</figcaption>
</figure>
</div>
</section>
<section id="aitchison-transformations-e.g.-clr" class="level2">
<h2 class="anchored" data-anchor-id="aitchison-transformations-e.g.-clr">Aitchison transformations (e.g.&nbsp;CLR)</h2>
<ul>
<li><p>Aitchison transformations are used to reduce compositional bias.</p></li>
<li><p>Balances, or ratios between taxa abundances, are conserved in compositional transformation: <span class="math inline">\(\frac{x}{y} = \frac{cx}{cy}\)</span></p></li>
</ul>
<p><img src="images/paste-3BA50E61.png" class="img-fluid"></p>
</section>
<section id="transformations" class="level2">
<h2 class="anchored" data-anchor-id="transformations">Transformations</h2>
<ul>
<li>Count data</li>
<li>Presence/absence</li>
<li>Compositional (percentages)</li>
<li><span class="math inline">\(Log_{10}\)</span></li>
<li><em>Aitchison</em> family of transformations (CLR, ALR, ILR)</li>
<li>Phylogenetic transformations (e.g.&nbsp;phILR)</li>
<li>Custom transformations</li>
</ul>
</section>
<section id="transformations-1" class="level2">
<h2 class="anchored" data-anchor-id="transformations-1">Transformations</h2>
<p>Create and add new assays in the data:</p>
<ul>
<li><a href="https://microbiome.github.io/OMA/exercises.html#transformations-1">OMA exercise 18.5</a></li>
</ul>
</section>
<section id="phylogenetic-transformations" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="phylogenetic-transformations">Phylogenetic transformations</h2>
<div>

</div>
<div class="quarto-layout-panel" data-layout="[[1,1], [1]]">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><em>- - there exists no single taxonomic resolution at which taxonomic variation unambiguously reflects functional variation, and at which environmental selection of certain functions - - unambiguously translates to a selection of specific taxa&nbsp;(Louca et al.&nbsp;2018).</em></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="images/paste-562EF513.png" class="img-fluid"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<p><img src="images/paste-8DB25B71.png" class="img-fluid"></p>
</div>
</div>
</div>
</section>
<section id="phylogenetic-balances-philr-transformation" class="level2 scrollable">
<h2 class="scrollable anchored" data-anchor-id="phylogenetic-balances-philr-transformation">Phylogenetic balances: phILR transformation</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-0974625F.png" class="img-fluid figure-img"></p>
<figcaption>Silverman et al.&nbsp;2017</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p><img src="images/paste-89510684.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="phylogenetic-balances-as-features-for-fli-predictions" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="phylogenetic-balances-as-features-for-fli-predictions">Phylogenetic balances as features for FLI predictions</h2>
<p>(Based on phILR; Silverman et al.&nbsp;2017)</p>
<p>Pathways in representative bacterial genomes of Clostridium subclusters IV and XIVa indicated the presence of e.g., ethanol fermentation pathways → endogenous ethanol producers associated with fatty liver?</p>
<p>In addition to age and sex, the models included differences in 11 microbial groups from class Clostridia, mostly belonging to orders Lachnospirales and Oscillospirales. Previously NAFLD-associated Clostridia XIVa group members were detected. Two species in Clostridia IV group were not previously associated with fatty liver disease.</p>
<p>Key associations validated in another Finnish cohort (N=258).</p>
<div class="columns">
<div class="column" style="width:70%;">
<p><img src="images/paste-A0B9F136.png" class="img-fluid"></p>
</div><div class="column" style="width:30%;">
<p><img src="images/paste-A43CFEC6.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="optional-task" class="level2">
<h2 class="anchored" data-anchor-id="optional-task">Optional task</h2>
<ul>
<li>If the data has phylogenetic tree, perform the phILR transformation (e.g.&nbsp;<code>GlobalPatterns</code> data set in the <strong>mia</strong> R package)</li>
</ul>
</section>
<section id="further-tasks" class="level2">
<h2 class="anchored" data-anchor-id="further-tasks">Further tasks</h2>
<p>If you complete the task fast, check out other <a href="https://microbiome.github.io/OMA/exercises.html">OMA Exercises</a> on <strong>data containers</strong>.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>